### PR TITLE
BUG: Write out Name Objects as UTF-8

### PR DESCRIPTION
**Pull request by rtpg**
The PDF spec describes name objects as utf-8 in general (see Appendix H,  3.2.4 Name Objects), and anything in `latin-1` is going to have the same byte representation.

### CODE REVIEW

**Comment by MartinThoma** [comment]
This PR fixes a bug, right? Could you give an example PDF / a minimal script that shows the bug?

**Comment by rtpg** [comment]
I believe that this bug was hit on a private PDF, so it would be hard for me to provide a full test case... but I'll look into it. I can of course make a unit test on `NameObject` itself, would that be good enough for considering a merge?

**Comment by MartinThoma** [comment]
It's already good enough to consider a merge. But before I hit the button, I want to ensure this doesn't break anything else.

I just became the maintainer yesterday; I still need to get familiar with the details. To do so, I'm currently increasing the test coverage.

File: PyPDF2/generic.py [hunk 1]
```diff
@@ -471,7 +471,7 @@ class NameObject(str, PdfObject):
     surfix = b_("/")
 
     def writeToStream(self, stream, encryption_key):
-        stream.write(b_(self))
+        stream.write(self.encode('utf8'))
```
Comment (MartinThoma) [inline PyPDF2/generic.py:474]:
```suggestion
        stream.write(self.encode("utf8"))
```

Comment (MartinThoma) [inline PyPDF2/generic.py:474]:
```suggestion
        stream.write(self.encode("utf8"))
```

Comment (MartinThoma) [reply PyPDF2/generic.py:474]: Could you please merge that tiny change? It's less about that change and more about making sure CI runs.

Comment (rtpg) [reply PyPDF2/generic.py]: I pulled in master and accidentally ran `black` across the file... I saw a commit doing that before, not sure why there are so many changes. 

For running the CI, seems like I need an approval from you 
image

Comment (MartinThoma) [reply PyPDF2/generic.py]: Thank you! I've approved and it seems that the 3.x versions are fine, but for the moment we also want to support 2.x

**Comment by MartinThoma** [comment]
Btw: Making too many unrelated changes makes it way less likely that your PR gets merged any time soon. It makes it way more time-consuming for me to check and thus I then prioritize others which are shorter.

**Comment by rtpg** [comment]
yeah of course, I'm going to rebase this to clean this up. Just wanted to get CI working so I can get this all green in one go

File: PyPDF2/generic.py [hunk 2]
```diff
@@ -469,23 +478,22 @@ class NameObject(str, PdfObject):
     surfix = b_("/")
 
     def writeToStream(self, stream, encryption_key):
-        stream.write(b_(self))
+        stream.write(self.encode("utf8"))
```
Comment (MartinThoma) [inline PyPDF2/generic.py:481]: Is that the only change that is not just a formatting change?

File: PyPDF2/generic.py [hunk 3]
```diff
@@ -469,23 +478,22 @@ class NameObject(str, PdfObject):
...
+        )
+        if debug:
+            print(name)
         try:
-            try:
-                ret=name.decode('utf-8')
-            except (UnicodeEncodeError, UnicodeDecodeError) as e:
-                ret=name.decode('gbk')
-            return NameObject(ret)
+            return NameObject(name.decode("utf-8"))
```
Comment (MartinThoma) [inline PyPDF2/generic.py:496]: Ah, here is another change :-)

Comment (pubpub-zz) [reply PyPDF2/generic.py:496]: Not sure this is matching table 3.3 rules...

**Comment by MartinThoma** [comment]
@rtpg Would you mind to update this PR? We did have a lot of changes recently in PyPDF2.

A test that fails with the current PyPDF2 and succeeds with your change would be amazing. This would help me a lot to understand why those changes are necessary.

**Comment by MartinThoma** [comment]
How is #651 related to #367?

File: PyPDF2/generic.py [hunk 4]
```diff
@@ -112,6 +112,7 @@ def readFromStream(stream):
         if nulltxt != b_("null"):
             raise utils.PdfReadError("Could not read Null object")
         return NullObject()
+
```
Comment (MartinThoma) [inline PyPDF2/generic.py:115]:
```suggestion
```

**Comment by MartinThoma** [comment]
I'm sorry, but I'm closing this PR now. It doesn't have any connected issue. If I execute the test

```python
def test_name_objects():
    # name objects get encoded via utf-8 (and not latin-1)
    name_object = NameObject("あいうえお")
    stream = BytesIO()
    name_object.write_to_stream(stream, None)
    assert stream.getvalue() == "あいうえお".encode("utf-8")
```

with the current `main` branch, it succeeds. So I assume that the issue got addressed elsewhere.

Thank you for the effort you put into it @rtpg :heart:

**Pull request closed without merging**